<!DOCTYPE html>
<html>

  <head>
    <meta charset='utf-8' />
    <meta http-equiv="X-UA-Compatible" content="chrome=1" />
    <meta name="description" content="Papacode : " />

    <link rel="stylesheet" type="text/css" media="screen" href="stylesheets/stylesheet.css">

    <title>Papacode</title>
  </head>

  <body>

    <!-- HEADER -->
    <div id="header_wrap" class="outer">
        <header class="inner">
          <a id="forkme_banner" href="https://github.com/vanSchaening/PapaCode">View on GitHub</a>

          <h1 id="project_title">Papacode</h1>
          <h2 id="project_tagline"></h2>

            <section id="downloads">
              <a class="zip_download_link" href="https://github.com/vanSchaening/PapaCode/zipball/master">Download this project as a .zip file</a>
              <a class="tar_download_link" href="https://github.com/vanSchaening/PapaCode/tarball/master">Download this project as a tar.gz file</a>
            </section>
        </header>
    </div>

    <!-- MAIN CONTENT -->
    <div id="main_content_wrap" class="outer">
      <section id="main_content" class="inner">
        <h1>
<a name="code-for-the-lionfish-project-" class="anchor" href="#code-for-the-lionfish-project-"><span class="octicon octicon-link"></span></a>Code for the Lionfish project </h1>

<p>This repository contains the scripts our lab has written for manipulating the output files from several bioinformatics platforms, mainly from STACKS. Most of the scripts either reconstruct short genome sequences or generate pseudogenomes based on the given input, returning them as .fasta files.</p>

<p>The kinds of pseudogenomes used by our group in this project are the following:
1. Concatenations of the sequences for an individual's two alleles for a particular loci.
2. Instead of concatenating the alleles, using IUPAC ambiguity codes in the SNP positions.
3. For an individual, concatenating all of the SNPs for selected loci.</p>

<p>Pseudomaker</p>

<p>Magic</p>

<p>Export To Fasta</p>

<p>input:  export data from stacks (usually a file called export_haps.tsv), and optionally a list of loci to filter by.
output: sequences reconstructed based on requests. It can either reconstruct biallelic versions of a locus, with both alleles concatenated, or monoallelic versions using IUPAC ambiguity codes. Instead of reconstructing the whole locus, it can instead make a string of concatenated SNPs.</p>
      </section>
    </div>

    <!-- FOOTER  -->
    <div id="footer_wrap" class="outer">
      <footer class="inner">
        <p class="copyright">Papacode maintained by <a href="https://github.com/vanSchaening">vanSchaening</a></p>
        <p>Published with <a href="http://pages.github.com">GitHub Pages</a></p>
      </footer>
    </div>

    

  </body>
</html>
